### PR TITLE
Add 'enable_on_boot' feature to power_supply

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

Add `enable_at_startup` option to the `power_supply` component. This is useful for boards that need a pin in a specific mode and state early in the boot process.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3411

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
power_supply:
  - id: atx_power_supply
    enable_time: 20ms
    keep_on_time: 10s
    enable_at_startup: true
    pin:
      number: 13
      inverted: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by kroimon — APPROVED** [review]
LGTM

File: esphome/components/power_supply/__init__.py [hunk 1]
```diff
@@ -18,6 +20,7 @@
         cv.Optional(
             CONF_KEEP_ON_TIME, default="10s"
         ): cv.positive_time_period_milliseconds,
+        cv.Optional(CONF_ENABLE_AT_STARTUP, default=False): cv.boolean,
```
Comment (jesserockz) [inline esphome/components/power_supply/__init__.py]: Can we change this to `ENABLE_ON_BOOT` to match wifi and ble flags?

File: esphome/components/power_supply/power_supply.h [hunk 2]
```diff
@@ -13,6 +13,7 @@ class PowerSupply : public Component {
   void set_pin(GPIOPin *pin) { pin_ = pin; }
   void set_enable_time(uint32_t enable_time) { enable_time_ = enable_time; }
   void set_keep_on_time(uint32_t keep_on_time) { keep_on_time_ = keep_on_time; }
+  void set_enable_at_startup(uint32_t enable_at_startup) { enable_at_startup_ = enable_at_startup; }
```
Comment (jesserockz) [inline esphome/components/power_supply/power_supply.h]:
```suggestion
  void set_enable_at_startup(bool enable_at_startup) { enable_at_startup_ = enable_at_startup; }
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Generally looks good, just the rename please

**Review by jesserockz — APPROVED** [review]
Thanks!

**Pull request merged**